consent-ui/src/app/ais/login/login.component.spec.ts

```diff
@@ -1,42 +1,109 @@
-import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import {async, ComponentFixture, TestBed} from '@angular/core/testing';
+import {ReactiveFormsModule} from '@angular/forms';
+import {LoginComponent} from './login.component';
+import {RouterTestingModule} from '@angular/router/testing';
+import {ActivatedRoute, ActivatedRouteSnapshot, Params} from '@angular/router';
+import {HttpClientTestingModule} from '@angular/common/http/testing';
+import {Observable, of} from 'rxjs';
+import {HttpHeaders, HttpResponse} from '@angular/common/http';
+import {AuthService} from '../../common/auth.service';
 
-import { ReactiveFormsModule } from '@angular/forms';
-import { LoginComponent } from './login.component';
-import { RouterTestingModule } from '@angular/router/testing';
-import { ActivatedRoute } from '@angular/router';
-import { StubUtilTests } from '../common/stub-util-tests';
-import { HttpClientTestingModule } from '@angular/common/http/testing';
+export class MockActivatedRoute {
+  snapshot: ActivatedRouteSnapshot;
+  param: Observable<Params>
+}
 
 describe('LoginComponent', () => {
   let component: LoginComponent;
   let fixture: ComponentFixture<LoginComponent>;
+  let route;
+  let authServiceSpy;
+  let authService: AuthService;
 
   beforeEach(async(() => {
+    route = new MockActivatedRoute();
+    route.snapshot = {
+      queryParams: {redirectCode: 'redirectCode654'},
+      parent: { params: {authId: 'authIdHere'} }
+    };
+
     TestBed.configureTestingModule({
       declarations: [ LoginComponent ],
       imports: [ReactiveFormsModule, HttpClientTestingModule, RouterTestingModule],
       providers: [
-        {
-          provide: ActivatedRoute,
-          useValue: {
-            snapshot: {
-              parent: { params: { authId: StubUtilTests.AUTH_ID } },
-              queryParams: { redirectCode: StubUtilTests.REDIRECT_ID }
-            }
-          }
-        }
+        { provide: ActivatedRoute, useValue: route }
       ]
     })
     .compileComponents();
   }));
 
   beforeEach(() => {
     fixture = TestBed.createComponent(LoginComponent);
+    authService = fixture.debugElement.injector.get(AuthService);
     component = fixture.componentInstance;
     fixture.detectChanges();
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should be true if the form is invalid', () => {
+    let headersOpt = new HttpHeaders({ 'Location': 'httpw://localhost:9876/?id=77168991' });
```

Comment:
The lines
```
 let headersOpt = new HttpHeaders({ 'Location': 'httpw://localhost:9876/?id=77168991' });
    let response = new HttpResponse({ body: { xsrfToken: 'tokenHere' }, headers: headersOpt, status: 200, statusText: 'geht' });
```
are pretty much same as
https://github.com/adorsys/open-banking-gateway/pull/638/files#diff-536917417e3e4dea0363f3f727c98904R81
and
https://github.com/adorsys/open-banking-gateway/pull/638/files#diff-536917417e3e4dea0363f3f727c98904R98
instead of duplicating code - they should be extracted in some function